File: TIC_TAC_TOE/TicTacToe.java
import java.util.*;

enum Symbol {
    X,
    O
}

class Player {
    String name;
    Symbol symbol;

    public Player(String name, Symbol symbol) {
        this.name = name;
        this.symbol = symbol;
    }
}

class Board {
    int size;
    Symbol[][] grid;
    int filledCells;

    public Board(int size) {
        this.size = size;
        this.grid = new Symbol[size][size];
        this.filledCells = 0;
    }

    public boolean isValidMove(int row, int col) {
        return row >= 0 && row < size &&
               col >= 0 && col < size &&
               grid[row][col] == null;
    }

    public boolean placeMove(int row, int col, Symbol symbol) {
        if(!isValidMove(row, col))
            return false;

        grid[row][col] = symbol;
        filledCells++;
        return true;
    }

    public boolean isFull() {
        return filledCells == size * size;
    }
}

class TicTacToe {

    Board board;
    Player player1;
    Player player2;
    Player currentPlayer;
    Player winner;
    boolean gameOver;

    public TicTacToe(int size, Player player1, Player player2) {
        board = new Board(size);
        this.player1 = player1;
        this.player2 = player2;
        currentPlayer = player1;
        winner = null;
        gameOver = false;
    }

    public boolean makeMove(int row, int col) {

        if(gameOver)
            return false;

        if(!board.placeMove(row, col, currentPlayer.symbol))
            return false;

        if(checkWinner(row, col)) {
            winner = currentPlayer;
            gameOver = true;
            return true;
        }

        if(board.isFull()) {
            gameOver = true;
            return true;
        }

        currentPlayer = (currentPlayer == player1)
                        ? player2
                        : player1;

        return true;
    }

    private boolean checkWinner(int row, int col) {

        Symbol symbol = currentPlayer.symbol;
        int n = board.size;

        boolean rowWin = true;
        for(int j = 0; j < n; j++) {
            if(board.grid[row][j] != symbol) {
                rowWin = false;
                break;
            }
        }

        boolean colWin = true;
        for(int i = 0; i < n; i++) {
            if(board.grid[i][col] != symbol) {
                colWin = false;
                break;
            }
        }

        boolean diagWin = true;
        if(row == col) {
            for(int i = 0; i < n; i++) {
                if(board.grid[i][i] != symbol) {
                    diagWin = false;
                    break;
                }
            }
        } else {
            diagWin = false;
        }

        boolean antiDiagWin = true;
        if(row + col == n - 1) {
            for(int i = 0; i < n; i++) {
                if(board.grid[i][n - 1 - i] != symbol) {
                    antiDiagWin = false;
                    break;
                }
            }
        } else {
            antiDiagWin = false;
        }

        return rowWin || colWin || diagWin || antiDiagWin;
    }

    public Player getWinner() {
        return winner;
    }

    public boolean isGameOver() {
        return gameOver;
    }

    public Player getCurrentPlayer() {
        return currentPlayer;
    }
}

File: TIC_TAC_TOE/TicTacToe.py
from enum import Enum


class Symbol(Enum):
    X = "X"
    O = "O"


class Player:
    def __init__(self, name: str, symbol: Symbol):
        self.name = name
        self.symbol = symbol


class Board:
    def __init__(self, size: int):
        self.size = size
        self.grid = [[None] * size for _ in range(size)]
        self.filled_cells = 0

    def isValidMove(self, row: int, col: int) -> bool:
        return (0 <= row < self.size and
                0 <= col < self.size and
                self.grid[row][col] is None)

    def placeMove(self, row: int, col: int, symbol: Symbol) -> bool:
        if not self.isValidMove(row, col):
            return False

        self.grid[row][col] = symbol
        self.filled_cells += 1
        return True

    def isFull(self) -> bool:
        return self.filled_cells == self.size * self.size


class TicTacToe:
    def __init__(self, size: int, player1: Player, player2: Player):
        self.board = Board(size)
        self.player1 = player1
        self.player2 = player2

        self.current_player = player1
        self.winner = None
        self.game_over = False

    def makeMove(self, row: int, col: int) -> bool:

        if self.game_over:
            return False

        if not self.board.placeMove(
                row,
                col,
                self.current_player.symbol):
            return False

        if self._checkWinner(row, col):
            self.winner = self.current_player
            self.game_over = True
            return True

        if self.board.isFull():
            self.game_over = True
            return True

        self.current_player = (
            self.player2
            if self.current_player == self.player1
            else self.player1
        )

        return True

    def _checkWinner(self, row: int, col: int) -> bool:

        symbol = self.current_player.symbol
        n = self.board.size

        row_win = all(
            self.board.grid[row][j] == symbol
            for j in range(n)
        )

        col_win = all(
            self.board.grid[i][col] == symbol
            for i in range(n)
        )

        diag_win = (
            row == col and
            all(
                self.board.grid[i][i] == symbol
                for i in range(n)
            )
        )

        anti_diag_win = (
            row + col == n - 1 and
            all(
                self.board.grid[i][n - 1 - i] == symbol
                for i in range(n)
            )
        )

        return row_win or col_win or diag_win or anti_diag_win

    def getWinner(self):
        return self.winner

    def isGameOver(self) -> bool:
        return self.game_over

    def getCurrentPlayer(self):
        return self.current_player